desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
+
+        # TODO: https://github.com/f0uriest/interpax/issues/19
+        integrand = Akima1DInterpolator(zeta, integrand, axis=-1).c
+
+        integrand = jnp.moveaxis(integrand, 1, -1)
+        assert integrand.shape == (4, pitch.size * M * L, N)
+        # For this algorithm, computing integrals via differences of primitives
+        # is preferable to any numerical quadrature. For example, even if the
+        # intersection points were evenly spaced, a composite Simpson's quadrature
+        # would require computing the spline on 1.8x more knots for the same accuracy.
+        primitive = polyval(intersect, polyint(integrand)).reshape(
+            pitch.size * M * L, N * R
+        )
+        sums = jnp.cumsum(
+            # Periodic boundary to compute bounce integrals of particles
+            # trapped outside this snapshot of the field lines.
+            jnp.diff(primitive, axis=-1, append=primitive[..., 0, jnp.newaxis])
+            # Multiply by mask that is false at knots of piecewise spline
+            # to avoid adding difference between primitives of splines at knots.
+            * jnp.append(jnp.arange(1, N * R) % R != 0, True),
+            axis=-1,
+        )
+        return fun((sums, is_bp)).reshape(pitch.size, M, L, N * R // 2)
+
+    def _quad_sin(name, pitch=None):
+        """Compute the bounce integral of the named quantity.
+
+        Parameters
+        ----------
+        name : ndarray
+            Name of quantity in ``data_index`` to compute the bounce integral of.
+        pitch : ndarray
+            λ values to evaluate the bounce integral at.
+            If None, uses the values given to the parent function.
+
+        Returns
+        -------
+        F : ndarray, shape(pitch, alpha, rho, (resolution - 1) * 2)
+            The last axis iterates through every bounce integral performed
+            along that field line padded by zeros.
+
+        """
+        pitch, intersect, _, bp1, bp2 = _compute_bp_if_given_pitch(
+            pitch, zeta, poly_B, poly_B_z, *original, err=True
+        )
+        intersect = intersect.reshape(pitch.size * M * L, N * NUM_ROOTS)
+        bp1 = fun((intersect, bp1))
+        bp2 = fun((intersect, bp2))
+        X = x * (bp2 - bp1)[..., jnp.newaxis] + bp2[..., jnp.newaxis]
+        assert X.shape == (pitch.size * M * L, N * 2, x.size)
+
+        def body(i, integral):
+            k = i % (N * 2)
+            j = i // (N * 2)
+            p = i // (M * L * N * 2)
+            v = j % pitch.size
+            # TODO: Add Hermite spline to interpax to pass in B_z[i].
+            integrand = interp1d(X[j, k], zeta, f[v]) / (
+                interp1d(X[j, k], zeta, B_sup_z[v])
+                * jnp.sqrt(1 - pitch[p] * interp1d(X[j, k], zeta, B[v]))
+            )
+            integral = put(integral, i, jnp.sum(w * integrand))
+            return integral
+
+        f = eq.compute(name, grid=grid, override_grid=False, data=data)[name].reshape(
```

Comment:
I think we need to be careful with an API like this. `override_grid=False` means that integral quantities like iota will be wrong, which will then mess with bounce integral

Reply:
I don't think the automatic selection of grids in eq.compute is currently jit compatible -- the grid.unique_idx will be empty arrays. The input dictionary data already stores things like `iota`

Reply:
How is the input `data` being computed though?

I see two main use cases:
1. computing stuff for plotting/analysis etc in a notebook, using `eq.compute` or something similar. In this case the quantities and grids are arbitrary so it's hard to guard against mistakes, but this is generally not compiled so the extra grid logic is fine and ensures quantities are accurate
2. Inside an objective function, where things are compiled but we know something about the quantities being computed and the grid (even if just the rho values) ahead of time, so we can manually precompute some stuff on the correct grid and then prime the data dict.

In either case I think it's best to leave the guards in place, but have them detect if the stuff that needs a special grid is already provided (which I think it currently doesn't but is an easy fix).

Reply:
Ok I'll remove the `override_grid=False` then. The input data is computed a few lines above here: https://github.com/PlasmaControl/DESC/pull/854/files#r1505087076. I'll need to make a new function like `_meshgrid_compress` to do the downsample/upsample between grids without the `grid.unique_idx`.